### ISSUE TITLE
[DOCKER-CRIU]  VM crash when trying to restore checkpoint.

### ISSUE THREAD

**Issue by jmserre** [xemul/criu#271]
(Hope I am at the right place, if wrong please help me re-direct to proper place)

Experimenting with Docker-CRIU.  Goal is to checkpoint container that takes a long time to start/load, so that can restore (in service faster)

criu version 2.9 (update 2.10 now)
Docker version 1.13.0-dev, build 67f607a, experimental
CentOS 7
uname -a
..... 4.9.0-1.el7.elrepo.x86_64 #1 SMP Sun Dec 11 15:43:54 EST 2016 x86_64 x86_64 x86_64 GNU/Linux
(can checkpoint restore the looper example)

(make sure container works well after it has finish loading all its data)

docker checkpoint create <name> toto
/var/lib/docker/containers/<containerId>/checkpoints/toto/criu.work
tail dump.log

```
(00.345743) cg: Writing CG image
(00.345868) sk unix: Dumping external sockets
(00.345875) Writing image inventory (version 1)
(00.346019) Running post-dump scripts
(00.346026)     RPC
(00.346259) Unfreezing tasks into 2
(00.346270)     Unseizing 3283 into 2
(00.346297)     Unseizing 3399 into 2
(00.373936) Writing stats
(00.373995) Dumping finished successfully
```

docker start --checkpoint toto containerName
VM dies nearly instantly
I do not have control of the VM, but the IT guys tell me on the VM console they see stuff like

[xxxxxxx.yyyyyyy]blk_update_request: I/O error, dev df0, sector 0
[zzzzzzz.vvvvvvv]blk_update_request: I/O error, dev df0, sector 0
....
Now post VM crash

It is like a new version of the container was starting to be created.
in 
/var/lib/docker/containers/new containerId/checkpoints/toto/criu.work
there is a restore<date> directory with a restore.log
which is 8842 lines long...

(00.177209)      1: pr1 Read 7fff4ada6000 12 pages
(00.177212)      1: pr1 Read 7fff4adf9000 2 pages
(01.352056)      1: Error (criu/pagemap.c:435): Can't read async pr bytes
(01.367011) Error (criu/cr-restore.c:1018): 10168 killed by signal 9: Killed
(01.386949) mnt: Switching to new ns to clean ghosts
(01.387019) Error (criu/cr-restore.c:1895): Restoring FAILED

Just trying to restore on the same machine.
[restore.log.txt](https://github.com/xemul/criu/files/711151/restore.log.txt)

Doing something wrong?  Missing resources?

**Comment by xemul** [comment]
Hm... The messages in VM logs show that there's error reading data from disk :\
Can you type 'dmesg' from inside VM and show the result?

Also, if I give you a patch, would you be able to test one?

**Comment by jmserre** [comment]
[dmseg.txt](https://github.com/xemul/criu/files/711406/dmseg.txt)

First thank you for helping and taking the time to respond.
As for a patch, I may need additional instructions, but will do my best to try it. I am no Linux guru but I have access to Linux experts in-house.

This seems relevant.

[   22.709604] nf_conntrack version 0.5.0 (65536 buckets, 262144 max)
[28845.623630] blk_update_request: I/O error, dev fd0, sector 0
[28845.623885] floppy: error -5 while reading block 0

(floppy drive????)

attached the full dmesg output.
(Another possibility, that the process of restoring image takes a lot more disk space than the original needed?)

**Comment by xemul** [comment]
Yes, floppy is really strange. Can you check where the image files are (likely in /var/lib/docker/containers/new containerId/checkpoints/toto/) and find images with names 'pages-*.img' then check they are readable by md5sum-ing them?

**Comment by jmserre** [comment]
(Thanks again!)
so there is one new /var/lib/docker/containers/xxxxx   were I had the restore.log
so the original is 
/var/lib/docker/containers/7c41cec559e5203fed0147cc7316b7289fcd3da9535e8b68ae71cfb7435b7544/checkpoints/toto

```
 ls -lt
total 220
drwxr-xr-x 4 root root  4096 Jan 16 14:52 criu.work
-rw-r--r-- 1 root root  4963 Jan 16 14:51 cgroup.img
-rw-r--r-- 1 root root    42 Jan 16 14:51 inventory.img
-rw-r--r-- 1 root root    80 Jan 16 14:51 pipes.img
-rw-r--r-- 1 root root    82 Jan 16 14:51 reg-files.img
-rw-r--r-- 1 root root    82 Jan 16 14:51 ipcns-var-10.img
-rw-r--r-- 1 root root    34 Jan 16 14:51 utsns-11.img
-rw-r--r-- 1 root root  2724 Jan 16 14:51 mountpoints-12.img
-rw-r--r-- 1 root root    36 Jan 16 14:51 pstree.img
-rw-r--r-- 1 root root   385 Jan 16 14:51 tmpfs-dev-66.tar.gz.img
-rw-r--r-- 1 root root   355 Jan 16 14:51 tmpfs-dev-69.tar.gz.img
-rw-r--r-- 1 root root    18 Jan 16 14:51 fs-74.img
-rw-r--r-- 1 root root    34 Jan 16 14:51 ids-74.img
-rw-r--r-- 1 root root   436 Jan 16 14:51 mm-74.img
-rw-r--r-- 1 root root   752 Jan 16 14:51 sigacts-74.img
-rw-r--r-- 1 root root   809 Jan 16 14:51 core-74.img
-rw-r--r-- 1 root root    86 Jan 16 14:51 pagemap-74.img
-rw-r--r-- 1 root root 49152 Jan 16 14:51 pages-2.img
-rw-r--r-- 1 root root    44 Jan 16 14:51 fdinfo-3.img
-rw-r--r-- 1 root root   784 Jan 16 14:51 core-1.img
-rw-r--r-- 1 root root    18 Jan 16 14:51 fs-1.img
-rw-r--r-- 1 root root    34 Jan 16 14:51 ids-1.img
-rw-r--r-- 1 root root   436 Jan 16 14:51 mm-1.img
-rw-r--r-- 1 root root   825 Jan 16 14:51 sigacts-1.img
-rw-r--r-- 1 root root    86 Jan 16 14:51 pagemap-1.img
-rw-r--r-- 1 root root 57344 Jan 16 14:51 pages-1.img
-rw-r--r-- 1 root root    44 Jan 16 14:51 fdinfo-2.img
-rw-r--r-- 1 root root    22 Jan 16 14:51 pipes-data.img
-rw-r-xr-x 1 root root    49 Jan 16 14:51 descriptors.json
-rw-r--r-- 1 root root   137 Jan 16 14:51 config.json
```
ls pa*
pagemap-1.img  pagemap-74.img  pages-1.img  pages-2.img

 md5sum pages-1.img
9a02e74a22676ffd250197e6c87d4904  pages-1.img

**Comment by xemul** [comment]
How about pages-2.img?

**Comment by jmserre** [comment]
md5sum pages-2.img
d8a18ae9903ec4ba3403b3e8fc1886e0  pages-2.img

**Comment by xemul** [comment]
OK. So images are OK. Let's now try to find out what's going on.
Here's the patch: https://github.com/xemul/criu/commit/b0f1e66e50aed12cf1bc112e095c40fd4e492384
Would you apply one (or get the criu-dev-piovs branch from this repo), compile criu and re-run with it?

Another option is -- you upload your images somewhere (provided it doesn't contain any sensitive information!) and post a link here so we could try to restore from them in our nodes :)

**Comment by jmserre** [comment]
Ok I think I found a good clue. (sorry for cross-posting)

so 

/var/lib/docker/containers/7c41cec559e5203fed0147cc7316b7289fcd3da9535e8b68ae71cfb7435b7544/checkpoints/toto
Is where the checkpoint was made.

 ls -l page*
-rw-r--r-- 1 root root    86 Jan 16 14:51 pagemap-1.img
-rw-r--r-- 1 root root    86 Jan 16 14:51 pagemap-74.img
-rw-r--r-- 1 root root 57344 Jan 16 14:51 pages-1.img
-rw-r--r-- 1 root root 49152 Jan 16 14:51 pages-2.img


I think 

/var/lib/docker/containers/22e2a2f70b43257b43ee596438e6be211cf8c59b3c14060b1fb78bfca4274646/checkpoints/toto

would have been the new container (e.g. restored from checkpoint)
because I see 
/var/lib/docker/containers/22e2a2f70b43257b43ee596438e6be211cf8c59b3c14060b1fb78bfca4274646/checkpoints/toto/criu.work/restore-2017-01-16T19:18:53Z/restore,log

in 
/var/lib/docker/containers/22e2a2f70b43257b43ee596438e6be211cf8c59b3c14060b1fb78bfca4274646/checkpoints/toto

-rw-r--r-- 1 root root      28312 Jan 16 18:58 pagemap-1.img
-rw-r--r-- 1 root root 3175591936 Jan 16 18:58 pages-1.img

there is no pages-2.img and page-1.img is huge.... this does not look normal?

**Comment by jmserre** [comment]
As for running my image, not an obvious solution, (my company would not let me do that) and the image is dependent on some other stuff, e.g. not trivial to be able to run it .  But I hope we will find out that there is a newbie mistake in my set-up.

**Comment by xemul** [comment]
Hm... You know, looking at pages-2 size I think that there's a BUG in CRIU code :\ Would you try the proposed patch to check my guess?

**Comment by xemul** [comment]
And actually, having a single pages-1.img file matches what I've seen in logs :) So would you also md5sum this single file as well?

**Comment by jmserre** [comment]
Hi I do not know how to 'professionally' use your patch, but I can use vi and stick it there manually.  If not have been using make for like 15-20 years, but I think make clean will do... if not will find on google 

So I already did md5sum on the pages-1 and pages-2 in the 'source'  one where the checkpoint was created.

Where the restore failed and there is only pages-1.img it took 20-30sec because the file is large but
 md5sum pages-1.img
879e40e3f89b052168439c90e1d4b474  pages-1.img

**Comment by xemul** [comment]
OK, so all files are readable and VM IO errors are not related. Good :)

As far as patches and building is concerned -- where did you get criu binary for your existing experiments? :D

**Comment by jmserre** [comment]
make clean
make
ls -lt 
total 18984
-rwx------ 1 root root 4230608 Jan 18 14:26 criu

Now where did I get it is a good question, it took me several attempts to get a working combination where I could have everything

http://download.openvz.org/criu/criu-2.9.tar.bz2

BUt I remember having trouble with bz2 (I am on centOS) anyway I will crash my VM soon, and put the new restore.log as soon as IT reboot my VM (which is not under my control.....)

**Comment by xemul** [comment]
Ah, so you've actually _compiled_ it :) That's good! So to apply the patch just wget this link (https://github.com/xemul/criu/commit/b0f1e66e50aed12cf1bc112e095c40fd4e492384.patch) from github and then do the

    patch -p1 < file_with_patch

command.

**Comment by jmserre** [comment]
Sorry about the delay, nobody was around in IT for a while.  Here is the end of the restore.log file....
(I can post the full file if needed)

0.332803)      1: pr1 Read 7f5983905000 2 pages
(00.332807)      1: pr1 Read 7f5983908000 1 pages
(00.332811)      1: pr1 Read 7f5983909000 2 pages
(00.332815)      1: pr1 Read 7f598390b000 1 pages
(00.332818)      1: pr1 Read 7f598390c000 1 pages
(00.332822)      1: pr1 Read 7fff4ada6000 12 pages
(00.332826)      1: pr1 Read 7fff4adf9000 2 pages
(01.711089)      1: Error (criu/pagemap.c:436): Can't read async pr bytes (2147479552 / 3175591936 read, 0 off, 612 iovs)
(01.731151) Error (criu/cr-restore.c:1018): 23163 killed by signal 9: Killed
(01.747020) mnt: Switching to new ns to clean ghosts
(01.747057) Error (criu/cr-restore.c:1895): Restoring FAILED.

Now my VM is not huge in terms of disk space, so is the restore need something like all the image (around 3.3G) + memory dump? of Xg this maybe my problem.  The whole checkpoint directory in the original container directory is only 500K

but in the 'restoring' directory it is 3G and probably incomplete, I must have around 6G left

**Comment by jmserre** [comment]
Ok I am wrong, I have around 30G  for /var/lib/.......   (I just deleted some old container and saved 3G, but It means I still had 26-27G freed earlier this morning (my time, Standard Eastern)

**Comment by jmserre** [comment]
Just to be sure, I installed CRIU 2.10, I completely cleaned-up docker (e.g. not container, no volume....) so 
/var/lib/docker/container/ was empty

re-did the looper example.  Notice that the container was restored at the same place (e.g. after the restore the containerId was the same).

Started my 3 containers for my real app.  Created a checkpoint of the container I am interested in 'freezing'

Made sure everything in there looks ok....
/var/lib/docker/containers/83d0b8fe6faa0b040ac2a013b39fd54aeb6d52988fafdd7e3ff7af9dea4f0053/checkpoints/toto/criu.work

docker start --checkpoint  toto <container name>
VM crash
after IT restart....

the restore.log is in 

/var/lib/docker/containers/83d0b8fe6faa0b040ac2a013b39fd54aeb6d52988fafdd7e3ff7af9dea4f0053/checkpoints/toto/criu.work/restore-2017-01-20T14:54:34Z

at the end....


00.179643)      1: pr1 Read 7f4e40da1000 18 pages
(00.179647)      1: pr1 Read 7f4e40ef5000 48 pages
(00.179651)      1: pr1 Read 7f4e40f27000 2 pages
(00.179655)      1: pr1 Read 7f4e40f2a000 1 pages
(00.179658)      1: pr1 Read 7f4e40f2b000 2 pages
(00.179662)      1: pr1 Read 7f4e40f2d000 1 pages
(00.179666)      1: pr1 Read 7f4e40f2e000 1 pages
(00.179670)      1: pr1 Read 7ffd0a6a2000 12 pages
(00.179674)      1: pr1 Read 7ffd0a6be000 2 pages
(01.173993)      1: Error (criu/pagemap.c:412): Can't read async pr bytes (2147479552 / 2560978944 read, 0 off, 932 iovs)
(01.185185) Error (criu/cr-restore.c:1012): 23872 killed by signal 9: Killed
(01.201051) mnt: Switching to new ns to clean ghosts
(01.201086) Error (criu/cr-restore.c:1893): Restoring FAILED.

So exactly the same than before. Anything else I can try?

**Comment by xemul** [comment]
OK. So the preadv() for some reason limits the amount of read data to ~2Gig. That's ... valid, though not expected :) I'll try to cook the fixing patch soon. Thanks for the help!

**Comment by xemul** [comment]
@jmserre , here's the patch (https://github.com/xemul/criu/commit/ed55f97162403c76936a3283c017fa226bf444c1.patch) that should be applied on top of the first one. Would you please check whether it helps your problem?

**Comment by jmserre** [comment]
(As usual thanks for your time/support)
A bit of feedback, tried the patch, still crash, but nobody in IT to reboot. When I get the VM back, will make sure the patch was applied that criu was recompiled (pretty sure I did....), and if the error was the same than before. Hopefully will have news tomorrow morning.

**Comment by jmserre** [comment]
Update: something did change because the restore file is different, I hope it is enough for you to figure out what is going on.  I have also attached the patched version of pagemap.  

[restore.log.txt](https://github.com/xemul/criu/files/729752/restore.log.txt)
[pagemap.c.txt](https://github.com/xemul/criu/files/729757/pagemap.c.txt)

So is what am I trying to do too big?  (E.g. my container?)  The bad news for me is this is relatively small compare to the real ones I want to checkpoint.

**Comment by xemul** [comment]
Whee! @jmserre , the fix was correct, thank you :) Now CRIU was managed to restore the memory and next failure is with net namespace configuration:

    iptables: Bad rule (does a matching rule exist in that chain?).

The rule that we're trying to add is

    iptables -w -t filter -D INPUT --protocol tcp -m mark ! --mark 0xC114 --source 127.0.0.1 --sport 863
6 --destination 127.0.0.1 --dport 34223 -j DROP

The thing that happens here is -- CRIU tries to _unlock_ a TCP socket, that was previously locked. And for some reason there's no netns in your Docker container, so CRIU does per-connection locking with iptables. Can you show the dump.log, please?

**Comment by xemul** [comment]
And answering your question regarding whether your container was too big -- no, it's not :) It's just us not (yet) testing such huge things and thus catching bugs from time to time. So thank you for this bug report.

**Comment by jmserre** [comment]
Attaching dump.log (e.g. when the checkpoint was originally created 20-Jan).  Because it is a bit big, I zipped it first....
[dumpLog.zip](https://github.com/xemul/criu/files/730006/dumpLog.zip)

I am not sure what is netns, but as I explained before the container I am trying to freeze is talking to 2 other ones.  Now one possible hack would be for us to implement a special flag -Di.want.to.freeze.you=1 where the container loads its data, close all socket, wait to be frozen, put in the logs it is ready.  After we restore it  -Dyou.are.restored=1  and re-open needed socket.

**Comment by xemul** [comment]
OK, so what I see in the logs is that container shares net namespace with the host, so CRIU doesn't dump one, treats all the TCP sockets as belonging to host and locks them with iptables.

There are several ways to go from here :)

1. Find out why the container is such and whether this is really what you need. I'm not an expert on Docker and need help from @avagin and @boucher. Gents, what can be the reason for netns-less container?

2. Before restore you, @jmserre , re-create the needed netfilter rules by hands so that CRIU is able to :) remove them during restore. Actually, this is needed anyway if you do want your connections to live migrate safely.

2a. Substitute iptables command with /bin/true to fool CRIU and _pretend_ that connections locking went well.

3. Add a tempting --ignore-tcp flag to CRIU so that it just kills all the established sockets and stubs them with unconnected ones.

4. You add your -D flag to your app with which it will close all TCP sockets so that CRIU doesn't try to lock/unlock them.

**Comment by boucher** [comment]
Docker provides `--net=host` I think to make certain kinds of isolation a bit easier. For example, if all you want is to make dependencies work more easily, and otherwise have the process work as if its running directly on the host, you can supply a bunch of flags that turn off a bunch of the namespaces. This is obviously very use case specific, and probably its actually over-used.

Interestingly, when trying to reproduce #277 this morning I also ended up just crashing the VM, probably because of the same thing. 

We could at the Docker level enforce not allowing the checkpoint of a container with `--net=host` I guess. I'm all for `--ignore-tcp` as a CRIU flag (I actually build a custom version of CRIU that does a much hackier/worse version of this for our own use). But there's no way to provide custom flags to CRIU in docker/runc/containerd yet...

**Comment by xemul** [comment]
Ah, indeed. The --net=host thing :) Thanks, @boucher
OK, I'll try to cook yet another patch for this and post one here. It shouldn't take long.
As far as passing custom option is concerned -- yes, we all want this to get fixed too (issue #269 and request on the ML for /etc/criu.conf).

**Comment by xemul** [comment]
@jmserre , here's the patch (https://github.com/xemul/criu/commit/1d8da101369a58b7373e6d6592ac724072fad517.patch) that will ... well, _should_ :) make CRIU ignore TCP sockets and replace them with just closed ones. Would you please try one?

**Comment by jmserre** [comment]
Thanks to @boucher  and @xemul.
Now I understand.  This is good news for me.  I thought to keep things simple and run the minimum of containers in CRIU I would use the --net=host option.  Also has a developer it is convenient for me to have direct access to real IPs/real debug ports.  Bun in real deployment we never use this set-up.   So I will first try the patch.

But I have no problem trying this with a more orthodox settings and reporting back on that too.

**Comment by HelenHarman** [comment]
@xemul Thanks for the patch. What git branch/checkout do I need to be on to apply the patch? I keep getting:
> error: patch failed: criu/sk-inet.c:163
error: criu/sk-inet.c: patch does not apply

**Comment by xemul** [comment]
@HelenHarman , this is for CRIU @jmserre was using :) i.e. the 2.9 one. For your convenience you can take the criu-dev-piovs branch from my CRIU repo, it already contains all the patches mentioned in this issue.

**Comment by jmserre** [comment]
Sorry for slow response, took me a long time to get VM back (rebooted by IT).  Now this is about retrying the same container restore (e.g. the one that was running with  --net=host).   I see a different error at the end of the restore.log

(15.257544)      1: 	Restore: family 2 type 2 proto 17 port 38485 state 7 src_addr 0.0.0.0
(15.257563)      1: 275 restore sndbuf 212992 rcv buf 212992
(15.257570)      1: 	restore priority 0 for socket
(15.257575)      1: 	restore rcvlowat 1 for socket
(15.257580)      1: 	restore mark 0 for socket
(15.257586)      1: Found fd 271 (id pipe:[207004]) in inherit fd list (caller inherit_fd_resolve_clash)
(15.257591)      1: Inherit fd 271 moved to 277 to resolve clash
(15.257597)      1: 		Create fd for 271
(15.257601)      1: 	Restoring fd 272 (state -> create)
(15.298453)      1: Error (criu/files-reg.c:1516): File var/local/Nuance/system/diagnosticLogs/nvc-uimas1 has bad size 118780 (expect 2530)
(15.322264) Error (criu/cr-restore.c:1009): 15046 exited, status=1
(15.334013) mnt: Switching to new ns to clean ghosts
(15.334596) Error (criu/cr-restore.c:1893): Restoring FAILED.

Mean something? E,g, beside the obvious fact that is is complaining that the size is bigger than expected. I have zipped and attached the full restore.log just in case....
(But anyway I suspect this means you previous patch passed the hurdle of the sockets
[restore.log..zip](https://github.com/xemul/criu/files/733630/restore.log.zip)

**Comment by xemul** [comment]
OK, so now we have double restore issue here :)
Yes, the error is about bad file size. Since you've restored container once it continued doing something with the  files and, apparently, changed its size.
In order to get the 2nd restore working you'll have to provide _exactly_ the same filesystem for it. This can be done with overlayfs by creating an empty layer on top of the checkpointed one and restoring the container it _it_. For the 2nd (the 3rd, the 4th and all others) restored you'll have to provide yet another overlay mountpoint on top of the checkpointed FS. In other words -- you have to snapshot the FS together with the container FS. Maybe this can be done purely with docker commads, maybe we'd require some help from Docker side (@boucher).

**Comment by tswift242** [comment]
@xemul I think docker needs to add better support for managing a criu checkpoint and a docker image as a pair. I've given my thoughts about this here: https://github.com/docker/docker/issues/30443#issuecomment-275534126.

Currently, what you can do (and what I do) is to do a ```docker commit``` right before the ```docker checkpoint create```. However, since the container process will be unpaused in between those two commands, there's the chance the filesystem could change in between, and then you'll have the same error as above (it's unlikely but possible).

**Comment by xemul** [comment]
@tswift242 thanks for commenting :) By the time we get to the state you describe, can we call docker commit from the criu's post-dump hook? At this time processes are frozen and do not modify the filesystem.

(hooks: https://criu.org/Action_scripts)

**Comment by xemul** [comment]
For the record, issue for multiple restores is #167

**Comment by jmserre** [comment]
Thanks, for my point of view (because of VM crash) it looks like nothing was restored, but I understand we were partially successful so the underlying image was being changed.  So I will redo everything from scratch and retry (still with the e --net=host) just to make sure your latest patch work.  Now the two persons in IT most likely to reboot my VM seems to be off today, I may have to wait till Monday....

Sorry if I am going off topic, but in theory if I restore successfully can I create multiple containers from the original checkpoint?  So Ideally start container, wait for it to finish it's load data phase, checkpoint.   Restore a A,B,C  instance of the original?

**Comment by xemul** [comment]
@jmserre , does even _first_ restore causes crash afterwards?

In theory you _can_ create several containers from the single image. But it will requires some tune-ups :) E.g. split the FS-s and make them have unique IP addresses.

**Comment by jmserre** [comment]
I could not resist,  I flushed all existing containers, volumes.  Restarted everything form scratch, did a new checkpoint, tries to restore, VM crash!.   So I may only have my VM back Monday morning but I twill check the restore to see if it is still related to a file having an unexpected size.

**Comment by xemul** [comment]
@jmserre , ouch :( OK, looking forward for the logs and more info on this.

**Comment by jmserre** [comment]
Sorry busy day today.  So this is the restore from last Friday (zipped). Up to now we got 3 problems:
1- The memory issue (greater than 2Gb)  got a patch
2-  The --net=host  (still using it for now) got a patch
3-   multiple restore

I cleaned everything restarted container, and did checkpoint, restore crashed again, but no obvious error in the logs (from my point of view), this maybe similar to problem number 2. I will let you comment but as suggested in this thread should I try "docker commit" then checkpoint? set-up without net host?

[restore.Friday27.zip](https://github.com/xemul/criu/files/740288/restore.Friday27.zip)

**Comment by xemul** [comment]
@jmserre , I see no errors in logs at all, this is the log from successful restore. What do you mean by "restore crashed"?

> should I try "docker commit" then checkpoint?

If you want to do multiple restores, then yes, doing somthing with FS will be required :) But for single restore it's not needed.

> set-up without net host?

Well, with the patch for TCP connections drop it looks like you don't but let's first clarify why the restore crashes.

**Comment by jmserre** [comment]
So yes me I type docker start --checkpoint toto containerName and the terminal becomes irresponsive right away.  I did try with a second terminal on the same VM and it too blocks.  A Linux expert told me to look into a few place including

dmesg  and I still see...

[    6.001956] audit: type=1130 audit(1485543104.273:79): pid=1 uid=0 auid=4294967295 ses=4294967295 msg='unit=systemd-tmpfiles-setup comm="systemd" exe="/usr/lib/systemd/systemd" hostname=? addr=? terminal=? res=success'
[    6.066933] audit: type=1305 audit(1485543104.338:80): audit_backlog_limit=320 old=64 auid=4294967295 ses=4294967295 res=1
[    6.082918] audit: type=1305 audit(1485543104.354:81): audit_enabled=1 old=1 auid=4294967295 ses=4294967295 res=1
[    6.082940] audit: type=1305 audit(1485543104.354:82): audit_pid=1782 old=0 auid=4294967295 ses=4294967295 res=1
[    6.632979] NET: Registered protocol family 40
[    7.542194] IPv6: ADDRCONF(NETDEV_UP): ksdev0: link is not ready
[    8.141413] vmxnet3 0000:0b:00.0 ksdev0: intr type 3, mode 0, 9 vectors allocated
[    8.142383] vmxnet3 0000:0b:00.0 ksdev0: NIC Link is Up 10000 Mbps
[   21.390380] nf_conntrack version 0.5.0 (65536 buckets, 262144 max)
[30685.822761] blk_update_request: I/O error, dev fd0, sector 0
[30685.823011] floppy: error -5 while reading block 0
[118702.734722] blk_update_request: I/O error, dev fd0, sector 0
[118702.734909] floppy: error -5 while reading block 0
[202820.803199] blk_update_request: I/O error, dev fd0, sector 0
[202820.803394] floppy: error -5 while reading block 0
[288978.543328] blk_update_request: I/O error, dev fd0, sector 0
[288978.543478] floppy: error -5 while reading block 0

This seems consistent to the message visible on the console before IT restart my VM.

**Comment by jmserre** [comment]
VM crash again. Only difference
<I did clean up all containers/volumes first>
docker-compose  -f ......
(Check everything works, container I want to freeze is up)
docker commit ..... (this is new)
docker checkpoint create ..... toto
(as fast as I can type it)
docker start --checkpoint toto ......
<VM dies>
So the commit did not help

**Comment by jmserre** [comment]
This may not be super-useful but as far I know I did the restore (docker start --checkpoint) around 19:01 GMT
In  the /var/log/messages

Jan 31 18:55:01 mt-dmz-postgres03 systemd: Starting user-995.slice.
Jan 31 18:55:01 mt-dmz-postgres03 systemd: Started Session 1073 of user pcp.
Jan 31 18:55:01 mt-dmz-postgres03 systemd: Starting Session 1073 of user pcp.
Jan 31 18:55:01 mt-dmz-postgres03 systemd: Removed slice user-995.slice.
Jan 31 18:55:01 mt-dmz-postgres03 systemd: Stopping user-995.slice.
Jan 31 18:57:05 mt-dmz-postgres03 dockerd: time="2017-01-31T18:57:05.383507621Z" level=error msg="Handler for DELETE /v1.25/containers/c592c8a
c4e22 returned error: No such container: c592c8ac4e22"
Jan 31 18:57:39 mt-dmz-postgres03 kernel: cgroup: new mount options do not match the existing superblock, will be ignored
Jan 31 18:58:01 mt-dmz-postgres03 systemd: Created slice user-995.slice.
Jan 31 18:58:01 mt-dmz-postgres03 systemd: Starting user-995.slice.
Jan 31 18:58:01 mt-dmz-postgres03 systemd: Started Session 1074 of user pcp.
Jan 31 18:58:01 mt-dmz-postgres03 systemd: Starting Session 1074 of user pcp.
Jan 31 18:58:01 mt-dmz-postgres03 systemd: Removed slice user-995.slice.
Jan 31 18:58:01 mt-dmz-postgres03 systemd: Stopping user-995.slice.
Jan 31 18:58:13 mt-dmz-postgres03 kernel: cgroup: new mount options do not match the existing superblock, will be ignored
Jan 31 18:58:13 mt-dmz-postgres03 kernel: ip6_tables: (C) 2000-2006 Netfilter Core Team
Jan 31 19:00:01 mt-dmz-postgres03 systemd: Started Session 1075 of user root.
Jan 31 19:00:01 mt-dmz-postgres03 systemd: Starting Session 1075 of user root.
Jan 31 19:01:01 mt-dmz-postgres03 systemd: Started Session 1076 of user root.
Jan 31 19:01:01 mt-dmz-postgres03 systemd: Starting Session 1076 of user root.
Jan 31 19:04:25 mt-dmz-postgres03 rsyslogd: [origin software="rsyslogd" swVersion="7.4.7" x-pid="2490" x-info="http://www.rsyslog.com"] start
Jan 31 19:04:07 mt-dmz-postgres03 kernel: Linux version 4.9.0-1.el7.elrepo.x86_64 (mockbuild@Build64R7) (gcc version 4.8.5 20150623 (Red Hat 4
.8.5-11) (GCC) ) #1 SMP Sun Dec 11 15:43:54 EST 2016

The 19:04:25 entry is the IT guy rebooting,.
I am not sure I understand the DELETE error, I did commit the wrong container first so I did a docker rmi on the image.

**Comment by xemul** [comment]
@jmserre , AFAIU your setup is -- a host, a docker in it and a VM inside the docker container. Am I correct? If yes, the dmesg and /var/log/messages you show here -- are they from the host or are they from inside the VM?

**Comment by xemul** [comment]
Also, would you pick up the https://github.com/xemul/criu/commit/08acbd7b7833f0c2b6b58274b093f24fb259184c.patch patch (or the criu-dev-piovs branch) and try to restore again? One thing in the restoring log bothers me and I'd like to clarify it with additional debug.

**Comment by jmserre** [comment]
Ok for the patch, will redo everything from scratch.

I may need to talk to IT if you want exact details, but
Bare metal server, CentOS VM, run docker in it, start my containers then use the docker commands directly from the OS.  In other words I am not trying to run docker from inside a container.

Another small detail I've switched to CRIU 2.10 in the middle of this long thread

**Comment by xemul** [comment]
@jmserre , thanks for helping with it :) BTW, 2.10 doesn't have the fix for big images reading :(

Also that's very useful to know that what reboots is the VM _inside_ which you do the restore. And that's really strange. Can you also check how much RAM your VM has and how much RAM your docker container uses. Maybe we hit OOM on restore and it goes killing the whole thing?

**Comment by jmserre** [comment]
Yes but my 2.10 as all the patches we discussed in this thread includeing the pagemap one, I just applied the other one you recommend. I think I have plenty of RAM 

KiB Mem : 82491040 total, 81605376 free,   358536 used,   527124 buff/cache
KiB Swap:  3141628 total,  3141628 free,        0 used. 81379024 avail Mem

The disk space is a little tighter around 30G free.  In theory the 3 containers image/volumes use together 10G so I still have 20G left.  But that was all the Hard drive left on that host.   I have to admit I wonder if I should not move to a bigger VM (e.g. more disk space) just to be sure, but it will take time to get from IT and then to set up everything again.  As I am sure you figured out this is a side exploratory project so also I try to hide from my boss how much time I am really spending on this, so trying to stay low-key
Will report on latest try soon.

**Comment by xemul** [comment]
82 GiGs of RAM and less than 4 Gig of container. Yes, this is pretty lots of memory :)

**Comment by jmserre** [comment]
Sorry for delays, I did it several hours ago, VM crash (well crash or lost all networking to it, etc.) IT just rebooted it now.  Did not see anything super obvious in dmesg, var/log.  Did ask local linux expert (same guy installed kernel to take a look)

[dumpRestore.zip](https://github.com/xemul/criu/files/745494/dumpRestore.zip)

The restore log seems similar (e.g. looks like it worked)
so just to make sure

1-remove all older container version/images (except official image I am using)
2-apply your latest patch (make clean, make, )
3- start 3 containers (docker-compose --
4-docker commit on the one I want to checkpoint
5-docker checkpoint create 'container-name'  toto
6-make sure only 2 container left running (docker ps)
7-docker start --checkpoint toto 'container-name'
terminal loose connection, can never log back to VM untill IT reboot's it.

If nothing new, I think tomorrow I will try to run it using normal docker overlay IP stuff (e.g. no --net=host)

**Comment by xemul** [comment]
Hmm... Yes, the restore.log looks correct and the data from pages.img should have reached its proper destinations. The reason for VM to crash/stop/stuck is indeed unknown. Looking forward to the experiment with another network mode.

Cc @avagin , maybe he has some ideas what can be checked.

**Comment by xemul** [comment]
@jmserre , I've double-checked (even triple-checked) the fix for 2Gb memory issue and now 99.9% sure that the memory is restored correctly. So the connectivity loss is somewhere else.

You told, that IT stuff reboots the VM. Do they do it only upon your requests? Is it possible to get from the KVM/Qemu-logs for this VM?

**Comment by jmserre** [comment]
I have good news for you.  Bad news for me.

I was able to restore my container!!!!!  But to do that I had to super-simplify what I am really trying to do, so I start only the single container I am interested in, because he cannot talk to his 'friends'  let it start empty (about 24 seconds) freeze it and restore it.  When I restore it the logs keep being filled as if nothing had happened. The restore took several seconds, but I suppose this is normal.

So I need to modify my docker-compose to get better networking between my 3 guys (but without --net=host) so I can really test what I am interested in.  So will not close this yet, will keep posted of my progress.

**Comment by jmserre** [comment]
Thank you very much for your help, I managed to do something better today (Still need to work on networking my containers) but I managed to save my target container after it did its real work of loading a model, and I was able to freeeze/unfreeze.

Now I have a million docker related question (can I restore the same checkpoint several times to create several copies, on a different server?).  I think eventually I will have to run after  @boucher 

But this morning I will 'bask' in my success.

Merci! Thank You! spasibo 
спасибо
